pubsub/src/manager.rs

```diff
@@ -0,0 +1,115 @@
+//! Provides an executor for subscription Futures.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::sync::{
+	atomic::{self, AtomicUsize},
+	Arc,
+};
+
+use crate::core::futures::sync::oneshot;
+use crate::core::futures::{future, Future};
+use crate::{
+	typed::{Sink, Subscriber},
+	SubscriptionId,
+};
+use log::{error, warn};
+use parking_lot::Mutex;
+
+/// Alias for an implementation of `futures::future::Executor`.
+pub type TaskExecutor = Arc<dyn future::Executor<Box<dyn Future<Item = (), Error = ()> + Send>> + Send + Sync>;
+
+/// Trait used to provide unique subscription ids.
+pub trait IdProvider {
+	// TODO: Maybe have this impl Into<u64>?
+	type Id: Clone + Default + Eq + Hash;
+
+	/// Returns next id for the subscription.
+	fn next_id(&self) -> Self::Id;
+}
+
+/// Trait used to drive subscription Futures to completion.
+pub trait SubscriptionManager {
```

Comment:
I'm not sure the trait is really needed, I'd rather have `Manager` with a default generic parameter of `u64` subscription `IdProvider`. Mainly to avoid importing extra traits to call functions.
Do you have some use cases in mind where having it behind trait would be useful?

Reply:
Yeah, my reasoning kinda was that if people wanted a "default" implementation of a `SubscriptionManager` they'd use the `Manager` struct, and if they wanted to roll their own they just had to conform to the `SubscriptionManager` trait. Is it reasonable to expect that people will want to use something other than the "default" implementation? If not, then yeah we don't need to the trait.

Reply:
Since the trait is not really used in the library at all I think a concrete implementation is enough. If someone is building their own solution and needs that abstraction they can introduce a trait themselves and implement it for the type from our library.